integrations: Add ClickUp integration.

PieterCK opened the pull request:
### Overview
Creates an incoming webhook integration for ClickUp. The main use case is getting notifications when new ClickUp items such as task, list, folder, space, goals are created, updated, or deleted.

Fixes zulip#26529.
CZO [thread](https://chat.zulip.org/#narrow/stream/127-integrations/topic/Fixing.20issue.3A.20ClickUp.20integration.20.2326529)

### Screenshots and screen captures:

<details>
<summary>Screenshots</summary>

Doc:
![image](https://github.com/user-attachments/assets/752992f2-b0c4-42a1-9675-83f28cf27ed3)

Generate URL for an integration menu:
![image](https://github.com/user-attachments/assets/26f63460-2cd5-4a03-9516-25ab100cb705)

- Task Created
![image](https://github.com/zulip/zulip/assets/108744694/393a7125-8668-4871-a689-847593d67070)
- Task Deleted
![image](https://github.com/zulip/zulip/assets/108744694/45aa87a6-84c8-4b40-8168-b31de55c9872)
- Task Updated: Rename Task
![image](https://github.com/zulip/zulip/assets/108744694/5ac0f1b4-4f74-4fab-91e5-7b09ba364c9a)
- Task Updated: Priority Change
![image](https://github.com/zulip/zulip/assets/108744694/1d4f0042-fa44-4646-8363-6422a1155fd8)
- Task Updated: Status Change
![image](https://github.com/zulip/zulip/assets/108744694/2bc4a986-9ec2-454a-93aa-3126c13b4d5b)
- Task Updated: Add Asignee
![image](https://github.com/zulip/zulip/assets/108744694/254d3791-bfa8-45c2-981c-4a508e6fac45)
- Updated: Time Spent Change
![image](https://github.com/zulip/zulip/assets/108744694/b6887157-27f3-42a3-967a-6636cf87efd3)
- Task Updated: Due Date Change
![image](https://github.com/user-attachments/assets/11a50498-b285-4278-b65d-70d2573dde49)
- Task Updated: New Comment
![image](https://github.com/zulip/zulip/assets/108744694/da3d1ef1-a1b7-4db6-a4b0-363883cabd46)
- Task Updated: Task Moved
![image](https://github.com/zulip/zulip/assets/108744694/633b73b5-9e55-4098-a740-3467153fbdaf)
- List Created
![image](https://github.com/zulip/zulip/assets/108744694/31186809-f747-44a6-9dde-3027de2cf362)
- List Deleted
![image](https://github.com/zulip/zulip/assets/108744694/d37c5676-f2c0-4bee-9edf-2e5260bea850)
- List Updated: Rename List
![image](https://github.com/zulip/zulip/assets/108744694/f27e643f-4bb0-4abe-b075-447e3fa1c1e5)
- Space Created
![image](https://github.com/zulip/zulip/assets/108744694/f822332e-f2ae-4adc-b46a-82b41efcf242)
- Space Updated: Rename Space
![image](https://github.com/zulip/zulip/assets/108744694/40e176c3-1ecd-43ec-bfc4-681845e13e88)
</details>

- `zulip_clickup.py` walkthrough:
[Screencast from 08-19-2024 10:34:01 AM.webm](https://github.com/user-attachments/assets/5010c7c2-1ab0-437c-841e-9b02ce0ee007)

PR for the integration script: [PR#824](https://github.com/zulip/python-zulip-api/pull/824/files)
### Testing
To use the `/devtools/integrations` for testing this integration, use this link:

```
/api/v1/external/clickup?api_key=<YOUR_BOT_API_KEY>&clickup_api_key=37621629_ca2a4e8fb484b4761af7692ea345863b83987cec1da23deef899f03b699877f1&team_id=25567147
```
It is prefilled with the necessary `clickup_api_key` & `team_id` variables. Note that when using this specific URL, you won't be able to test the folder and goal-related test fixtures and only task, list and space-related test fixtures.
<details>


<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK committed:
integrations: Add ClickUp integration.

Creates an incoming webhook integration for ClickUp.
The main use case is getting notifications when new ClickUp
items such as task, list, folder, space, goals are created,
updated or deleted.

Fixes zulip#26529.

timabbott commented on zerver/lib/integrations.py:
```diff
@@ -8,7 +8,6 @@
 from django.utils.translation import gettext_lazy
 from django_stubs_ext import StrPromise
 from typing_extensions import TypeAlias
-
```
This looks like a typo/mistake.

PieterCK replied:
noted

timabbott commented on zerver/webhooks/clickup/api_endpoints.py:
```diff
@@ -0,0 +1,53 @@
+    
+    try:
+        response = ClickUpSession(headers=headers).get(
+            "https://api.clickup.com/api/v2/" + path,
```
Use an f-string here, or perhaps better, `urllib` helpers.

timabbott commented on zerver/webhooks/clickup/api_endpoints.py:
```diff
@@ -0,0 +1,53 @@
+        super().__init__(role="clickup", timeout=5, **kwargs)
+
+def make_clickup_request(path:str, api_key: str) -> dict:
+    headers: Dict[str, str] = {
```
From a security standpoint, it would be good to have validation that `path` contains only expected characters for an ID -- not `/`s, for example, but an allowlist seems more correct than a blocklist.

PieterCK replied:
I added a function to help verify the URL path but coulnd't push it in this commit. I'll update it in the next commit

timabbott commented on zerver/webhooks/clickup/view.py:
```diff
@@ -0,0 +1,223 @@
+    elif UPDATED in event_code:
+        body = generate_updated_event_message(item_data=item_data,payload=payload, event_item_type=event_item_type)
+
+    return topic, body
```
Your editor seems to be misconfigured -- none of these files seem to have newlines at their ends.

timabbott commented on zerver/webhooks/clickup/view.py:
```diff
@@ -0,0 +1,223 @@
+    event_code = payload["event"].tame(check_string)
+    topic = "ClickUp Notification"
+    
+    event_item_type = event_code.replace(CREATED, "").replace(UPDATED, "").replace(DELETED, "")
```
This probably deserves a comment, since it's a bit of a weird thing to do.

timabbott commented on zerver/webhooks/clickup/view.py:
```diff
@@ -0,0 +1,223 @@
+        new_time_spent =  convert_unixdatetime(int(raw_time_spent), format="%H:%M:%S")
+        return (
+            "\n~~~ quote\n :stopwatch: Time spent changed to **{new_value}**\n~~~\n"
+        ).format(new_value = new_time_spent)
```
This probably needs a final `else` clause?

timabbott commented on zerver/webhooks/clickup/view.py:
```diff
@@ -0,0 +1,223 @@
+def generate_delete_event_message(event_item_type: str) -> str:
+    return (
+        "\n:trash_can:  A {event_item_type} has been deleted from your ClickUp space!"
+    ).format(event_item_type=event_item_type.title())
```
Are the double-spaces in these messages intentional?

PieterCK commented:
Update: added more tests to 100% coverage.
 
@timabbott 
Hi Tim, hope you're not too busy with GSOC stuff and other issues. Would you mind giving a quick follow up my PR? Thank you!

laurynmm commented on zerver/webhooks/clickup/doc.md:
```diff
@@ -0,0 +1,56 @@
+
+1. Follow the instructions given by the script.
+
+    **Note**: keep watch for your browser, you will be redirected to ClickUp authorization page
```
I'd revise this to be:

"Follow the instructions in the terminal and keep an eye on your browser as you will be redirected to a ClickUp authorization page."

laurynmm reviewed (commented):
@PieterCK - Thanks for all the work on this integration! I did a review of the documentation part of these changes only.

Let me know if you have any questions about my feedback comments. Overall it looks pretty good and follows our new structure for integration docs.

One general change I'd make is to reorder the instructions to follow an order more like the one we use in the [Trello integration doc](https://zulip.com/integrations/doc/trello):
- Set up the Zulip bot/URL
- Get the ClickUp information
- Download the python script and run it

And as far as formatting in these files, we usually leave an empty line after headers and include brackets in the documentation pages, so instead of ...

```
![](image_location)
# Header
* text
{tip}
```

it would be ...

```
![](image_location)

# Header

* text

{tip}
```

PieterCK commented on zerver/webhooks/clickup/doc.md:
```diff
@@ -0,0 +1,56 @@
+
+
+1. Collect your ClickUp **Client ID & Client Secret** :
+    1.  Go to <https://app.clickup.com/settings/team/clickup-api> and click **Create an App** button.
```
Yeah, I agree with you here. But, it seems like there's a bug when using different types of lists (ordered & unordered) inside the {start_tabs}/{end_tabs} block. 

Here's what is rendered when I try to use the * unordered list:
<details>

![image](https://github.com/zulip/zulip/assets/108744694/9c46e486-45f1-47d5-8b28-48ca4d94f8d2)
![image](https://github.com/zulip/zulip/assets/108744694/1631fe77-1735-4053-8871-59c66523d371)
using the - unordered list yields the same result

</details>

Using unordered list outside and inside the block seems to change all other list type inside the block into unordered list: 
<details>

![image](https://github.com/zulip/zulip/assets/108744694/5045d9f0-c559-4b25-b704-c45dd5c3fc56)
![image](https://github.com/zulip/zulip/assets/108744694/94021325-40d6-4cc2-b0e4-fa11765ec333)

</details>

The best I could think of it to "manually" use bolded hyphens to act as the unordered list:
<details>

![image](https://github.com/zulip/zulip/assets/108744694/274a7187-ddf4-429f-9e21-37415632ae22)
![image](https://github.com/zulip/zulip/assets/108744694/979b7746-f65d-4b65-ab48-ba16de2c8c0d)

</details>

laurynmm replied:
Ah! Okay. I removed a CSS rule recently from `integrations.css` that seems to be coming in play here. We need to add a rule for unordered list elements inside an ordered list element.

@karlstolley mentioning you here to see if my suggestion below seems like the correct fix here.

@PieterCK - You could do this update as a separate prep-commit here ... or maybe a separate PR would be better. Because it would be good to fix this sooner rather than later and I'm not sure where in the review process the overall new ClickUp integration is.

```scss
.portico-landing.integrations {
...
    & ol li {
        list-style: none;

        & ul li {
            list-style: inherit;
        }
    }
```

test in airbrake doc

```markdown
1. {!generate-webhook-url-basic.md!}

    * test one
    * two test
    * test three
```

With the CSS changes that renders as ...

<details>
<summary>Test screenshot</summary>

*Note the unordered list element at the bottom is rendered as expected as well.*
![Screenshot from 2024-04-10 16-00-06](https://github.com/zulip/zulip/assets/63245456/72efac9f-d855-4f49-9eb2-bcc16317f16a)
</details>

PieterCK replied:
noted, I'll shoot up a new PR for the `integrations.css` fix

PieterCK replied:
tested it with ClickUp doc, much nicer :)

![image](https://github.com/zulip/zulip/assets/108744694/a1fbf348-19d0-416b-b764-8370463a3441)

PieterCK replied:
created a new issue just to track the PR, #29669 
and the PR for the css fix: #29671
@laurynmm

PieterCK commented on zerver/webhooks/clickup/doc.md:
```diff
@@ -0,0 +1,56 @@
+    1.  Go to <https://app.clickup.com/settings/team/clickup-api> and click **Create an App** button.
+
+    1.  You will be prompted for Redirect URL(s). Enter your zulip app URL.
+        e.g. `YourZulipApp.com`.
```
Noted. But, I was only able to find {{ api_url }} to generate the server URL. The problem is that it also includes "/api" right after the base URL, which I'm afraid will redirect the user to the API doc. 
![image](https://github.com/zulip/zulip/assets/108744694/8522dcc9-fbbd-446b-82cd-932f9850aa32)

laurynmm replied:
I think we pass `zulip_url` to the context for documentation pages as well and that shouldn't have the `/api` added. What happens if you try that?

PieterCK replied:
ahh gotcha, works like a charm. Thanks! Tried `base_url` and just `url` but to no avail earlier XD

laurynmm replied:
The context for documentation pages is built in `zerver/views/documentation.py` for future reference, which is where I spotted it.

PieterCK commented:
@laurynmm thanks for the feedbacks! Here's what the doc currently looks like:

![image](https://github.com/zulip/zulip/assets/108744694/82ca0630-f54b-446a-b81f-c12011dbb573)
If there are any more changes you'd like to suggest, feel free to leave another comment!

alya commented:
@PieterCK Looks like there are some test failures to fix for this PR to be ready for the next round of review.

PieterCK commented:
> @PieterCK Looks like there are some test failures to fix for this PR to be ready for the next round of review.

Fixed it. Looking forward for your review :+1:

laurynmm reviewed (commented):
@PieterCK - I did another pass on the documentation updates for this new integration. Let me know if you have any questions!

@timabbott - Let me know if you want me to do a review of the integration implementation.

PieterCK commented on zerver/webhooks/clickup/__init__.py:
```diff
@@ -0,0 +1,39 @@
+        return [item.value for item in cls]
+
+
+class EventItemType(ConstantVariable):
```
Constant variables are declared in `clickup/__init__.py` and not in `view.py` to avoid circular import because `api_endpoints.py` also uses a common constant variable; `EventItemType`.

PieterCK commented:
@kennethnrk Hello Kenneth, could you please review this PR? Thank you!

@zulipbot label "buddy review"

kennethnrk commented:
@PieterCK I'm unable to manually test this integration, I keep getting a response saying `bad request`.
![image](https://github.com/zulip/zulip/assets/29176437/d4b7cc55-6605-4032-b881-7c4704af5bb2)

Is it because I don't have a valid clickup `team_id`?

PieterCK commented:
Yeah, this integration is a bit different from the other ones because ClickUp doesn't provide complete human-readable data in most of their payload. 

To make things worse, they don't have a user-friendly UI or workflow to set up webhooks. Instead, you have to make API calls to their API endpoints to register the webhook integration ( Zulip endpoint URL & which events). To get the API keys to access their APIs, you must go through their [OAuth flow](https://clickup.com/api/developer-portal/authentication). 

This is the purpose of the helper script I submitted at this [PR](https://github.com/zulip/python-zulip-api/pull/824). It lives in the python-zulip-api repo and is required to integrate Zulip with ClickUp. For reference, our [Trello integration](https://zulip.com/integrations/doc/trello) follows a similar procedure to be integrated, and I largely followed that process for this script.

In addition to registering our webhook endpoint with ClickUp, this script has another function: appending a couple of variables to the user's webhook URL query. One of these variables is the `team_id` that you mentioned, and the other is the ClickUp API key I mentioned earlier. These variables will be used in the main Zulip codebase to do callbacks to ClickUp API endpoints and get additional context for the webhook payloads. 

That is why our current tool wasn't able to test the integration because the webhook URL it uses doesn't have the `team_id` and `clickup_api_key` as expected. Even if it has those the integration would attempt to call ClickUp API endpoints using dummy credentials (team_id & API key) and would fail. To do manual test, I think you have to actually connect the webhook to the ClickUp using tools like [ultrahook](https://www.ultrahook.com/)...

This whole PR is actually the first PR that I've submitted to Zulip, it is quite hefty and complicated. If I were to re-do it, I'd split it into a few commits the first one being a very rudimentary integration without callbacks and thus simpler notifications just so that it is easier to be tested and partially merged, then add the more complicated features like callbacks and detailed notifications in separate commits following that. This is also a suggestion my mentor @sbansal1999 endorsed. I think I will do it, but given the cost of such refactoring and the priority of this PR, this task would find itself near the bottom of my TODO list...

For now, I don't think you have to go through all that hassle to test the integration. It would already be very helpful if I could get a second pair of eyes to review the code styling and overall logic of the PR. @kennethnrk  @sbansal1999 what do you guys think?

PieterCK commented:
@zulipbot remove "maintainer review"

PieterCK commented:
@zulipbot remove "documentation review"

PieterCK commented:
Update: 
- Cleared up labels
- Added prefilled webhook URL with all the necessary variables for testing purposes in `devtools/integrations`
- Updated some of the test fixtures
- Updated the PR to follow the recent changes in the typing module (switching to builtin types)

kennethnrk commented:
Glad we can finally test this using the dev environment. LGTM.

@zulipbot remove "buddy review"

PieterCK commented:
@sbansal1999 this one's ready for your review too

PieterCK commented:
Update: 
- Refactored the main endpoint code and tests. I didn't include any git diff because this PR hasn't gotten a lot of code review, but let me know if that would help with reviewing.
- Refactored the docs to reflect the recent changes to the integration script in [#824](https://github.com/zulip/python-zulip-api/pull/824)
- Updated the docs screenshot & the screen record in the PRs description.

Hi, @laurynmm could you please take a look at this when you have time? It has only passed `buddy review` and received some doc reviews from you. Let me know if you need me to split the PR into more commits or if you need help with testing, I'd be happy to make this PR easier to review :)

PieterCK commented:
@zulipbot remove "mentor review"

alya commented:
@laurynmm tagging you for the next review round on this one.

PieterCK commented:
Update:
- rebased the PR to resolve the merge conflict.
- updated the integration to use the `config_options` framework for inputting the API credentials.

alya commented:
@Niloth-p are you up for doing the next round of review on this one?
